File: shared/sim/weapons.js
// Projectile simulation — shared by the SP client and the Go server.
// Owns the pooled projectile state and steps motion / missile guidance /
// collisions, applying gameplay outcomes (hp, dead, ttl) and returning a list
// of events for the caller to turn into FX + sound. No THREE meshes, no DOM.
import { Vector3 } from './vec.js';

const _tmp = new Vector3();

export class Projectiles {
  constructor(K) {
    this.K = K;                 // constants.weapons
    this.max = K.max;
    this.pos = [];
    this.vel = [];
    this.ttl = new Float32Array(this.max);
    this.age = new Float32Array(this.max);
    this.team = new Array(this.max).fill(null);
    this.kind = new Array(this.max).fill('bolt');
    this.target = new Array(this.max).fill(null);
    for (let i = 0; i < this.max; i++) { this.pos.push(new Vector3()); this.vel.push(new Vector3()); }
    this.cursor = 0;
  }

  spawn(pos, vel, team, ttl, target = null, kind = 'bolt') {
    const i = this.cursor;
    this.cursor = (this.cursor + 1) % this.max;
    this.pos[i].copy(pos);
    this.vel[i].copy(vel);
    this.ttl[i] = ttl;
    this.age[i] = 0;
    this.team[i] = team;
    this.kind[i] = kind;
    this.target[i] = target;
  }

  playerMissileActive() {
    for (let i = 0; i < this.max; i++) {
      if (this.ttl[i] > 0 && this.team[i] === 'player' && this.kind[i] === 'missile') return true;
    }
    return false;
  }

  playerMissileGuided() {
    for (let i = 0; i < this.max; i++) {
      if (this.ttl[i] > 0 && this.team[i] === 'player' && this.kind[i] === 'missile') {
        return !!(this.target[i] && !this.target[i].dead);
      }
    }
    return false;
  }

  // world: { player, enemies, pods, bonuses }  -> mutates state, returns events[]
  step(dt, world, K) {
    const { player, enemies, pods, bonuses } = world;
    const KW = this.K;
    const events = [];

    for (let i = 0; i < this.max; i++) {
      if (this.ttl[i] <= 0) continue;

      this.ttl[i] -= dt;
      this.age[i] += dt;
      const isMissile = this.kind[i] === 'missile';

      if (isMissile && this.team[i] === 'player') {
        const sp = this.vel[i].length();
        if (sp > 1e-3 && sp < KW.missileMaxSpeed) this.vel[i].multiplyScalar(1 + KW.missileSelfPropel * dt);
        if (this.age[i] > KW.missileGuideDelay) {
          const tgt = this.target[i];
          if (tgt && !tgt.dead) {
            const desired = _tmp.copy(tgt.position).sub(this.pos[i]).normalize().multiplyScalar(this.vel[i].length());
            this.vel[i].lerp(desired, 1 - Math.pow(KW.missileGuideRate, dt));
          }
        }
      }
      this.pos[i].addScaledVector(this.vel[i], dt);

      // --- collisions ---
      let hit = false;
      const dmg = isMissile ? KW.missileDmg : KW.cannonDmg;
      if (this.team[i] === 'player') {
        for (const e of enemies.list) {
          if (e.dead) continue;
          const r = e.radius + (isMissile ? KW.missileHitPad : 0);
          if (this.pos[i].distanceToSquared(e.position) < r * r) {
            e.hp -= dmg;
            e.flash = 1;
            events.push({ kind: 'enemyHit', pos: this.pos[i].clone(), isMissile });
            if (e.hp <= 0) {
              e.dead = true;
              events.push({ kind: 'enemyKill', pos: e.position.clone(), accent: e.stats.accent || 0xffcc55 });
            } else if (isMissile) {
              events.push({ kind: 'missileBurst', pos: this.pos[i].clone() });
            }
            hit = true;
            break;
          }
        }
        if (!hit && bonuses && bonuses.hitByShot(this.pos[i])) hit = true;
      } else {
        if (player.alive && this.pos[i].distanceToSquared(player.position) < KW.playerHitRadius * KW.playerHitRadius) {
          const absorbed = player.invuln > 0;   // grace period soaks it (no hurt sound)
          player.damage(KW.enemyDmgPlayer);
          events.push({ kind: 'playerHit', pos: this.pos[i].clone(), absorbed });
          hit = true;
        }
        if (!hit) {
          for (const pod of pods.list) {
            if (pod.dead) continue;
            if (this.pos[i].distanceToSquared(pod.position) < pod.radius * pod.radius) {
              pod.hp -= KW.enemyDmgPod;
              events.push({ kind: 'podHit', pos: this.pos[i].clone() });
              if (pod.hp <= 0) {
                pod.dead = true;
                events.push({ kind: 'podKill', pos: pod.position.clone(), pod });
              }
              hit = true;
              break;
            }
          }
        }
      }
      if (hit) this.ttl[i] = 0;
    }

    return events;
  }
}

File: server/game/weapons.go
package game

import "math"

// Pooled projectiles — the Go twin of shared/sim/weapons.js.

const (
	teamPlayer = "player"
	teamEnemy  = "enemy"
	kindBolt   = "bolt"
	kindMsl    = "missile"
)

type Projectiles struct {
	K      Block // constants.weapons
	Max    int
	Pos    []Vec3
	Vel    []Vec3
	Ttl    []float64
	Age    []float64
	Team   []string
	Kind   []string
	Owner  []string // player id that fired it ("" for enemy shots)
	Target []*Enemy
	Cursor int
}

func newProjectiles(kw Block) *Projectiles {
	n := int(kw["max"])
	return &Projectiles{
		K: kw, Max: n,
		Pos: make([]Vec3, n), Vel: make([]Vec3, n),
		Ttl: make([]float64, n), Age: make([]float64, n),
		Team: make([]string, n), Kind: make([]string, n),
		Owner:  make([]string, n),
		Target: make([]*Enemy, n),
	}
}

func (p *Projectiles) Spawn(pos, vel Vec3, team string, ttl float64, target *Enemy, kind, owner string) {
	i := p.Cursor
	p.Cursor = (p.Cursor + 1) % p.Max
	p.Pos[i] = pos
	p.Vel[i] = vel
	p.Ttl[i] = ttl
	p.Age[i] = 0
	p.Team[i] = team
	p.Kind[i] = kind
	p.Owner[i] = owner
	p.Target[i] = target
}

// PlayerMissileActive reports whether `owner` already has a guided missile in
// the air — the "one missile on screen" rule is per player, so in co-op one
// pilot's missile never blocks another's.
func (p *Projectiles) PlayerMissileActive(owner string) bool {
	for i := 0; i < p.Max; i++ {
		if p.Ttl[i] > 0 && p.Team[i] == teamPlayer && p.Kind[i] == kindMsl && p.Owner[i] == owner {
			return true
		}
	}
	return false
}

// Step advances the pool and resolves collisions, mutating enemy/pod/ship
// state. Returns events for the arena to broadcast.
func (p *Projectiles) Step(dt float64, w *World) []Event {
	kw := p.K
	var events []Event

	for i := 0; i < p.Max; i++ {
		if p.Ttl[i] <= 0 {
			continue
		}
		p.Ttl[i] -= dt
		p.Age[i] += dt
		isMissile := p.Kind[i] == kindMsl

		if isMissile && p.Team[i] == teamPlayer {
			sp := p.Vel[i].Length()
			if sp > 1e-3 && sp < kw["missileMaxSpeed"] {
				p.Vel[i].MultiplyScalar(1 + kw["missileSelfPropel"]*dt)
			}
			if p.Age[i] > kw["missileGuideDelay"] {
				tgt := p.Target[i]
				if tgt != nil && !tgt.Dead {
					desired := tgt.Position
					desired.Sub(p.Pos[i]).Normalize().MultiplyScalar(p.Vel[i].Length())
					p.Vel[i].Lerp(desired, 1-math.Pow(kw["missileGuideRate"], dt))
				}
			}
		}
		p.Pos[i].AddScaledVector(p.Vel[i], dt)

		hit := false
		dmg := kw["cannonDmg"]
		if isMissile {
			dmg = kw["missileDmg"]
		}

		if p.Team[i] == teamPlayer {
			for _, e := range w.Fleet.List {
				if e.Dead {
					continue
				}
				r := e.Radius
				if isMissile {
					r += kw["missileHitPad"]
				}
				if p.Pos[i].DistanceToSq(e.Position) < r*r {
					e.HP -= dmg
					e.Flash = 1
					events = append(events, Event{Kind: "enemyHit", Pos: p.Pos[i]})
					if e.HP <= 0 {
						e.Dead = true
						events = append(events, Event{Kind: "enemyKill", Pos: e.Position})
					} else if isMissile {
						events = append(events, Event{Kind: "missileBurst", Pos: p.Pos[i]})
					}
					hit = true
					break
				}
			}
			if !hit && bonusHitByShot(w.Bonuses, p.Pos[i]) {
				hit = true
			}
		} else {
			hr := kw["playerHitRadius"] * kw["playerHitRadius"]
			for _, sh := range w.shipList() {
				if sh.Alive && p.Pos[i].DistanceToSq(sh.Pos) < hr {
					absorbed := sh.Invuln > 0
					sh.Damage(kw["enemyDmgPlayer"])
					events = append(events, Event{Kind: "playerHit", Pos: p.Pos[i], Absorbed: absorbed})
					hit = true
					break
				}
			}
			if !hit {
				for _, pod := range w.Pods.List {
					if pod.Dead {
						continue
					}
					if p.Pos[i].DistanceToSq(pod.Position) < pod.Radius*pod.Radius {
						pod.HP -= kw["enemyDmgPod"]
						events = append(events, Event{Kind: "podHit", Pos: p.Pos[i]})
						if pod.HP <= 0 {
							pod.Dead = true
							events = append(events, Event{Kind: "podKill", Pos: pod.Position})
						}
						hit = true
						break
					}
				}
			}
		}
		if hit {
			p.Ttl[i] = 0
		}
	}
	return events
}
